### PR TITLE
Support SFT and DPO training for Qwen2VL

### PR DESCRIPTION
**Pull request by LiuXTao**
This PR supports training the multimodal model **Qwen2-VL**, with priority given to **SFT** and **DPO** training.

To run, you can refer to:

SFT script: `bash examples/vision_scripts/train_sft_qwen2vl.sh`

DPO script: `bash examples/vision_scripts/train_dpo_qwen2vl.sh`

I conducted training tests for SFT (trained for 3 epochs) and DPO (trained for 1 epoch). The training curves are as follows, where the **loss converges normally**, and metrics such as DPO's accuracy and reward increase as expected.

_sft curve_
image
_dpo curve_
image

To ensure training accuracy, I performed an **SFT comparison on Llama-Factory**. During training, I disabled data shuffling on both sides to ensure identical inputs and maintained the same learning rate. As a result, **the loss curve and grad_norm curve align perfectly**.
WeChatWorkScreenshot_2edd75cd-2076-4dd6-bef1-438907567202

This PR **already** supports:

- SFT and DPO training for Qwen2VL
- An option to freeze the vision model
- Support for passing mrope externally to Qwen2VL during training, avoiding issues with position ID misalignment

**Currently supporting / Will support** in the future:

- Enabling ring attention in the vision model
- Training for Intern2.5-VL

### CODE REVIEW

**Comment by nrailg** [comment]
Awesome! That's exactly what I need.

**Comment by hijkzzz** [comment]
Due to limited bandwidth, I briefly reviewed the content of the MR. This MR is excellent, but there are some issues we haven’t resolved yet:  

1. How to implement general multimodal support instead of supporting only a single model.  
2. The dataset part appears to have been copied from Llama Factory. How can we achieve general multimodal dataset support instead of relying on this fixed dataset?  

We haven’t figured out these two issues very clearly yet.

**Review by catqaq — COMMENTED** [review]
Good job, it's a great implementation! It is recommended to upgrade `get_qwen2_vl_utils` to a more general `get_vl_utils`, and remove any model-specific components from the main code. If there are any references in the data related code, please also indicate them appropriately.

**Comment by hijkzzz** [comment]
> > Due to limited bandwidth, I briefly reviewed the content of the MR. This MR is excellent, but there are some issues we haven’t resolved yet:
> > 
> > 1. How to implement general multimodal support instead of supporting only a single model qwen2-vl.
> > 2. The dataset part appears to have been copied from Llama Factory. How can we achieve general multimodal dataset support instead of relying on this hacked dataset?
> > 
> > We haven’t figured out these two issues very clearly yet.
> 
> @hijkzzz @catqaq Thank you for your review. Regarding your questions and suggestions, I have the following thoughts:
> 
> 1. For supporting more general models, I will add some general functions to reuse common logic as much as possible. For model-specific logic, I will use the model_arch parameter combined with if-statements to handle special cases for different models.
> 2. Regarding the dataset part, I did reference llama-factory and made some citations in certain places (I'm sorry if these citations are not very obvious or if I missed some places). I have considered making the dataset as general as possible by abstracting the load_dataset logic into blending_dataset() in openrlhf. However, the post-processing of image datasets is indeed quite specialized. If we aim for generalization, it might become a more complex implementation like llama-factory, which is one of the reasons I retained the post-processing logic. After thinking about it for the past few days, I think it can reuse structures similar to SFTDataset/RewardDataset in openrlhf, abstract the post-processing of multimodal datasets into class functions, and provide some switches to support differentiation.
> 
> Do you think these methods are appropriate?

Good idea! We hope to keep the code as simple and as general as possible, rather than hacking the code for quick implementation.

**Comment by LiuXTao** [comment]
> > > Due to limited bandwidth, I briefly reviewed the content of the MR. This MR is excellent, but there are some issues we haven’t resolved yet:
> > > 
> > > 1. How to implement general multimodal support instead of supporting only a single model qwen2-vl.
> > > 2. The dataset part appears to have been copied from Llama Factory. How can we achieve general multimodal dataset support instead of relying on this hacked dataset?
> > > 
> > > We haven’t figured out these two issues very clearly yet.
> > 
> > 
> > @hijkzzz @catqaq Thank you for your review. Regarding your questions and suggestions, I have the following thoughts:
> > 
> > 1. For supporting more general models, I will add some general functions to reuse common logic as much as possible. For model-specific logic, I will use the model_arch parameter combined with if-statements to handle special cases for different models.
> > 2. Regarding the dataset part, I did reference llama-factory and made some citations in certain places (I'm sorry if these citations are not very obvious or if I missed some places). I have considered making the dataset as general as possible by abstracting the load_dataset logic into blending_dataset() in openrlhf. However, the post-processing of image datasets is indeed quite specialized. If we aim for generalization, it might become a more complex implementation like llama-factory, which is one of the reasons I retained the post-processing logic. After thinking about it for the past few days, I think it can reuse structures similar to SFTDataset/RewardDataset in openrlhf, abstract the post-processing of multimodal datasets into class functions, and provide some switches to support differentiation.
> > 
> > Do you think these methods are appropriate?
> 
> Good idea! We hope to keep the code as simple and as general as possible, rather than hacking the code for quick implementation. Of course, this is quite challenging due to the poor code reusability of VLMs.

@hijkzzz Hi, I have reconstructed the data set and training code to improve the universality and usability of the data set code, reduce the redundancy in the training code, and support more models for greater scalability. Please review the code at your convenience. thank you!

The next step is to support ring attn and more models, such as intern2.5-vl.

**Comment by hijkzzz** [comment]
Thank you! I will continue the review when I have bandwidth.

**Comment by hijkzzz** [comment]
@khazic helps to review. thanks
@khazic suggests Openrlhf may have another repo for vision models as the poor code reusability of the vision models.
I'm also considering this. If we create another repository, this MR will be cherry-picked over.

**Comment by John-Ge** [comment]
https://github.com/TideDra/lmm-r1
This repo supports vision language model training using OpenRLHF. Maybe this does not require many modifications.

**Comment by chuangzhidan** [comment]
> This PR supports training the multimodal model **Qwen2-VL**, with priority given to **SFT** and **DPO** training.
> 
> To run, you can refer to:
> 
> SFT script: `bash examples/vision_scripts/train_sft_qwen2vl.sh`
> 
> DPO script: `bash examples/vision_scripts/train_dpo_qwen2vl.sh`
> 
> I conducted training tests for SFT (trained for 3 epochs) and DPO (trained for 1 epoch). The training curves are as follows, where the **loss converges normally**, and metrics such as DPO's accuracy and reward increase as expected.
> 
> _sft curve_ image _dpo curve_ image
> 
> To ensure training accuracy, I performed an **SFT comparison on Llama-Factory**. During training, I disabled data shuffling on both sides to ensure identical inputs and maintained the same learning rate. As a result, **the loss curve and grad_norm curve align perfectly**. WeChatWorkScreenshot_2edd75cd-2076-4dd6-bef1-438907567202
> 
> This PR **already** supports:
> 
> * SFT and DPO training for Qwen2VL
> * An option to freeze the vision model
> * Support for passing mrope externally to Qwen2VL during training, avoiding issues with position ID misalignment
> 
> **Currently supporting / Will support** in the future:
> 
> * Enabling ring attention in the vision model
> * Training for Intern2.5-VL

至今还没更新到分支上吗，看不到这2个sh脚本:
SFT script: bash examples/vision_scripts/train_sft_qwen2vl.sh
DPO script: bash examples/vision_scripts/train_dpo_qwen2vl.sh
想体验下

**Comment by chuangzhidan** [comment]
> This PR supports training the multimodal model **Qwen2-VL**, with priority given to **SFT** and **DPO** training.
> 
> To run, you can refer to:
> 
> SFT script: `bash examples/vision_scripts/train_sft_qwen2vl.sh`
> 
> DPO script: `bash examples/vision_scripts/train_dpo_qwen2vl.sh`
> 
> I conducted training tests for SFT (trained for 3 epochs) and DPO (trained for 1 epoch). The training curves are as follows, where the **loss converges normally**, and metrics such as DPO's accuracy and reward increase as expected.
> 
> _sft curve_ image _dpo curve_ image
> 
> To ensure training accuracy, I performed an **SFT comparison on Llama-Factory**. During training, I disabled data shuffling on both sides to ensure identical inputs and maintained the same learning rate. As a result, **the loss curve and grad_norm curve align perfectly**. WeChatWorkScreenshot_2edd75cd-2076-4dd6-bef1-438907567202
> 
> This PR **already** supports:
> 
> * SFT and DPO training for Qwen2VL
> * An option to freeze the vision model
> * Support for passing mrope externally to Qwen2VL during training, avoiding issues with position ID misalignment
> 
> **Currently supporting / Will support** in the future:
> 
> * Enabling ring attention in the vision model
> * Training for Intern2.5-VL

敢问下，使用的是哪个数据集
